examples/images_contours_and_fields/demo_bboximage.py

```diff
@@ -38,7 +38,7 @@
 a = np.vstack((a, a))
 
 # List of all colormaps; skip reversed colormaps.
-maps = sorted(m for m in plt.cm.cmap_d if not m.endswith("_r"))
+maps = sorted(m for m in plt.colormaps() if not m.endswith("_r"))
```

Comment:
I had the same thought: https://github.com/matplotlib/matplotlib/pull/16991#issuecomment-607548990

I'm not sure what the best step is to expose the access, but would agree that I don't want to have to go to pyplot if not necessary. Maybe rather than a `cmap_d`, we could do something like your MutableMapping suggestion and then override the get/set methods to ensure copy/immutable return.

Reply:
*Sorry if this is a bit unstructured. I just stated typing and ideas unfolded along the way.*

There would be two ways:

1. The `cm` module currently uses simple functions `get_cmap()` and `register_cmap()`. In that spirit, you could add a function `get_cmap_names()` (naming t.b.d.) but we have to make a distinction between `get_cmap` returning an instance and that function (probably?) returning strings.

2. The alternative would be to expose a `ColormapRegistry` object, which provides all necessary functionality. That would likely be subclassing `MutableMapping` to look dict-like but still perform all relevant checks (maybe it would alternatively only implement `Mapping` and provide an explicit register method if we feel the mutation should be more special (i.e. we don't want `del registry['jet']` or `registry['jet'] = Colormap(...)`).
This would also replace the module level functions; at least their implementation. Whether they will additionally stay for backward compatibility is a separate topic.

In the end 2 is the more complex approach. However, I have a slight preference for it because it properly encapsulates the colormap handling. On the downside it changes the API from calling simple cm functions to an object. But then again, we could move the actual instance to a variable `matplotlib.colormaps`, which may make more sense and users would usually not need to import matplotlib.cm anymore; and `pyplot.colormaps` could later even point to the same instance (with some API compatibility because the keys of that mapping would then colormap names, so `for name in plt.colormaps()` would still work.

Reply:
I agree with pretty much everything you're saying. I also don't think 2 would be that hard to implement. An issue with naming will be found with: #11413 which is already proposing to rename cm to colormaps. This discussion is really more a part of the MEP: https://matplotlib.org/3.2.1/devel/MEP/MEP21.html
Maybe we should do some larger work on an overarching vision there and then decide on how best to get there?

Reply:
If we provide colormap registry access from `matplotlib` directly, `cm.get_cmap` and `cm.register_cmap` become obsolete. There won't be a need for the end user to import `cm`, which in turn reduces the need for renaming (#11413). But we could still find other names for that.

It would be reasonable to detail the overall vision.